File: Repo/opencog/cogutil/c++/opencog-cogutil.cpp
/*
 * opencog-cogutil.cpp
 * 
 * OpenCog Cogutil - C++ Utility Library
 * A collection of utility functions and classes for OpenCog framework
 * 
 * This single-file implementation demonstrates core utility functionality
 * including logging, configuration management, and common data structures.
 */

#include <iostream>
#include <string>
#include <vector>
#include <map>
#include <memory>
#include <sstream>
#include <fstream>
#include <chrono>
#include <iomanip>
#include <cctype>

namespace opencog {
namespace cogutil {

// Logger class for debugging and informational output
class Logger {
public:
    enum Level {
        DEBUG,
        INFO,
        WARN,
        ERROR
    };

private:
    Level minLevel;
    std::ostream& output;

public:
    Logger(Level level = INFO, std::ostream& out = std::cout) 
        : minLevel(level), output(out) {}

    void log(Level level, const std::string& message) {
        if (level >= minLevel) {
            output << "[" << getCurrentTime() << "] " 
                   << levelToString(level) << ": " 
                   << message << std::endl;
        }
    }

    void debug(const std::string& msg) { log(DEBUG, msg); }
    void info(const std::string& msg) { log(INFO, msg); }
    void warn(const std::string& msg) { log(WARN, msg); }
    void error(const std::string& msg) { log(ERROR, msg); }

    void setLevel(Level level) { minLevel = level; }

private:
    std::string levelToString(Level level) const {
        switch(level) {
            case DEBUG: return "DEBUG";
            case INFO:  return "INFO";
            case WARN:  return "WARN";
            case ERROR: return "ERROR";
            default:    return "UNKNOWN";
        }
    }

    std::string getCurrentTime() const {
        auto now = std::chrono::system_clock::now();
        auto time = std::chrono::system_clock::to_time_t(now);
        std::stringstream ss;
        ss << std::put_time(std::localtime(&time), "%Y-%m-%d %H:%M:%S");
        return ss.str();
    }
};

// Configuration manager for storing key-value pairs
class Config {
private:
    std::map<std::string, std::string> data;

public:
    void set(const std::string& key, const std::string& value) {
        data[key] = value;
    }

    std::string get(const std::string& key, const std::string& defaultValue = "") const {
        auto it = data.find(key);
        return (it != data.end()) ? it->second : defaultValue;
    }

    bool has(const std::string& key) const {
        return data.find(key) != data.end();
    }

    bool loadFromFile(const std::string& filename) {
        std::ifstream file(filename);
        if (!file.is_open()) return false;

        std::string line;
        while (std::getline(file, line)) {
            // Skip comments and empty lines
            if (line.empty() || line[0] == '#') continue;

            size_t pos = line.find('=');
            if (pos != std::string::npos) {
                std::string key = line.substr(0, pos);
                std::string value = line.substr(pos + 1);
                // Trim whitespace
                key.erase(0, key.find_first_not_of(" \t"));
                key.erase(key.find_last_not_of(" \t") + 1);
                value.erase(0, value.find_first_not_of(" \t"));
                value.erase(value.find_last_not_of(" \t") + 1);
                set(key, value);
            }
        }
        return true;
    }

    void dump() const {
        for (const auto& pair : data) {
            std::cout << pair.first << " = " << pair.second << std::endl;
        }
    }
};

// Timer utility for performance measurement
class Timer {
private:
    std::chrono::high_resolution_clock::time_point startTime;
    std::string name;
    bool running;

public:
    Timer(const std::string& timerName = "Timer") 
        : name(timerName), running(false) {}

    void start() {
        startTime = std::chrono::high_resolution_clock::now();
        running = true;
    }

    double elapsed() const {
        if (!running) return 0.0;
        auto endTime = std::chrono::high_resolution_clock::now();
        std::chrono::duration<double> diff = endTime - startTime;
        return diff.count();
    }

    void stop(Logger* logger = nullptr) {
        if (running) {
            double time = elapsed();
            running = false;
            if (logger) {
                std::stringstream ss;
                ss << name << " completed in " << time << " seconds";
                logger->info(ss.str());
            }
        }
    }
};

// String utilities
class StringUtils {
public:
    static std::vector<std::string> split(const std::string& str, char delimiter) {
        std::vector<std::string> tokens;
        std::stringstream ss(str);
        std::string token;
        while (std::getline(ss, token, delimiter)) {
            if (!token.empty()) {
                tokens.push_back(token);
            }
        }
        return tokens;
    }

    static std::string join(const std::vector<std::string>& strings, const std::string& delimiter) {
        std::stringstream ss;
        for (size_t i = 0; i < strings.size(); ++i) {
            if (i > 0) ss << delimiter;
            ss << strings[i];
        }
        return ss.str();
    }

    static std::string toLower(const std::string& str) {
        std::string result = str;
        for (char& c : result) {
            c = std::tolower(c);
        }
        return result;
    }

    static std::string toUpper(const std::string& str) {
        std::string result = str;
        for (char& c : result) {
            c = std::toupper(c);
        }
        return result;
    }

    static std::string trim(const std::string& str) {
        size_t start = str.find_first_not_of(" \t\n\r");
        size_t end = str.find_last_not_of(" \t\n\r");
        return (start == std::string::npos) ? "" : str.substr(start, end - start + 1);
    }
};

} // namespace cogutil
} // namespace opencog

// Demonstration of cogutil functionality
int main() {
    using namespace opencog::cogutil;

    std::cout << "OpenCog Cogutil - C++ Utility Library Demo" << std::endl;
    std::cout << "===========================================" << std::endl << std::endl;

    // Logger demonstration
    Logger logger(Logger::INFO);
    logger.info("Cogutil library initialized");
    logger.debug("This debug message won't show (level too low)");
    logger.warn("This is a warning message");
    logger.error("This is an error message");

    logger.setLevel(Logger::DEBUG);
    logger.debug("Now debug messages are visible");
    std::cout << std::endl;

    // Config demonstration
    Config config;
    config.set("opencog.version", "1.0.0");
    config.set("atomspace.enabled", "true");
    config.set("cogserver.port", "17001");
    
    logger.info("Configuration loaded:");
    config.dump();
    std::cout << std::endl;

    logger.info("Retrieving specific config: port = " + 
                config.get("cogserver.port"));
    std::cout << std::endl;

    // Timer demonstration
    Timer timer("Processing");
    timer.start();
    
    logger.info("Simulating some work...");
    // Simulate work
    int sum = 0;
    for (int i = 0; i < 1000000; ++i) {
        sum += i;
    }
    
    timer.stop(&logger);
    std::cout << std::endl;

    // String utilities demonstration
    logger.info("String utilities demonstration:");
    std::string text = "OpenCog,AtomSpace,CogServer,Cogutil";
    auto parts = StringUtils::split(text, ',');
    
    logger.info("Split result:");
    for (const auto& part : parts) {
        std::cout << "  - " << part << std::endl;
    }

    std::string joined = StringUtils::join(parts, " + ");
    logger.info("Joined: " + joined);

    logger.info("Uppercase: " + StringUtils::toUpper("opencog rocks"));
    logger.info("Lowercase: " + StringUtils::toLower("OPENCOG ROCKS"));
    logger.info("Trimmed: '" + StringUtils::trim("  spaced out  ") + "'");

    std::cout << std::endl;
    logger.info("Cogutil demonstration complete!");

    return 0;
}

File: Repo/opencog/cogutil/Python/opencog-cogutil.py
#!/usr/bin/env python3
"""
opencog-cogutil.py

OpenCog Cogutil - Python Utility Library
A collection of utility functions and classes for OpenCog framework

This single-file implementation demonstrates Python's strengths:
- Dynamic typing and duck typing
- Rich standard library
- Decorator pattern for elegant code
- Context managers for resource management
"""

import sys
import time
import logging
from datetime import datetime
from typing import Any, Dict, List, Optional, Callable
from functools import wraps
from contextlib import contextmanager


class Logger:
    """
    Logger class showcasing Python's logging module integration
    Demonstrates: Standard library usage, singleton pattern
    """
    
    def __init__(self, name: str = "OpenCog", level: int = logging.INFO):
        self.logger = logging.getLogger(name)
        self.logger.setLevel(level)
        
        # Create console handler with formatting
        handler = logging.StreamHandler(sys.stdout)
        formatter = logging.Formatter(
            '[%(asctime)s] %(levelname)s: %(message)s',
            datefmt='%Y-%m-%d %H:%M:%S'
        )
        handler.setFormatter(formatter)
        self.logger.addHandler(handler)
    
    def debug(self, msg: str) -> None:
        self.logger.debug(msg)
    
    def info(self, msg: str) -> None:
        self.logger.info(msg)
    
    def warn(self, msg: str) -> None:
        self.logger.warning(msg)
    
    def error(self, msg: str) -> None:
        self.logger.error(msg)
    
    def set_level(self, level: int) -> None:
        self.logger.setLevel(level)


class Config:
    """
    Configuration manager using dictionary
    Demonstrates: Dictionary operations, file I/O, property access
    """
    
    def __init__(self):
        self._data: Dict[str, str] = {}
    
    def set(self, key: str, value: str) -> None:
        """Set a configuration value"""
        self._data[key] = value
    
    def get(self, key: str, default: str = "") -> str:
        """Get a configuration value with default"""
        return self._data.get(key, default)
    
    def has(self, key: str) -> bool:
        """Check if key exists"""
        return key in self._data
    
    def load_from_file(self, filename: str) -> bool:
        """Load configuration from file"""
        try:
            with open(filename, 'r') as f:
                for line in f:
                    line = line.strip()
                    # Skip comments and empty lines
                    if not line or line.startswith('#'):
                        continue
                    
                    if '=' in line:
                        key, value = line.split('=', 1)
                        self.set(key.strip(), value.strip())
            return True
        except FileNotFoundError:
            return False
    
    def dump(self) -> None:
        """Print all configuration"""
        for key, value in sorted(self._data.items()):
            print(f"{key} = {value}")
    
    def __getitem__(self, key: str) -> str:
        """Allow dictionary-style access"""
        return self._data[key]
    
    def __setitem__(self, key: str, value: str) -> None:
        """Allow dictionary-style setting"""
        self._data[key] = value


class Timer:
    """
    Timer utility using context manager
    Demonstrates: Context manager protocol, decorators
    """
    
    def __init__(self, name: str = "Timer", logger: Optional[Logger] = None):
        self.name = name
        self.logger = logger
        self.start_time: Optional[float] = None
        self.elapsed_time: Optional[float] = None
    
    def start(self) -> None:
        """Start the timer"""
        self.start_time = time.time()
    
    def stop(self) -> float:
        """Stop the timer and return elapsed time"""
        if self.start_time is None:
            return 0.0
        self.elapsed_time = time.time() - self.start_time
        if self.logger:
            self.logger.info(f"{self.name} completed in {self.elapsed_time:.6f} seconds")
        return self.elapsed_time
    
    def __enter__(self):
        """Context manager entry"""
        self.start()
        return self
    
    def __exit__(self, *args):
        """Context manager exit"""
        self.stop()


def timeit(func: Callable) -> Callable:
    """
    Decorator to time function execution
    Demonstrates: Python decorators, functional programming
    """
    @wraps(func)
    def wrapper(*args, **kwargs):
        start = time.time()
        result = func(*args, **kwargs)
        elapsed = time.time() - start
        print(f"[TIMER] {func.__name__} took {elapsed:.6f} seconds")
        return result
    return wrapper


class StringUtils:
    """
    String utilities showcasing Python's string methods
    Demonstrates: Static methods, string manipulation, list comprehensions
    """
    
    @staticmethod
    def split(text: str, delimiter: str = ',') -> List[str]:
        """Split string by delimiter, filtering empty strings"""
        return [s.strip() for s in text.split(delimiter) if s.strip()]
    
    @staticmethod
    def join(strings: List[str], delimiter: str = ',') -> str:
        """Join strings with delimiter"""
        return delimiter.join(strings)
    
    @staticmethod
    def to_lower(text: str) -> str:
        """Convert to lowercase"""
        return text.lower()
    
    @staticmethod
    def to_upper(text: str) -> str:
        """Convert to uppercase"""
        return text.upper()
    
    @staticmethod
    def trim(text: str) -> str:
        """Remove leading/trailing whitespace"""
        return text.strip()
    
    @staticmethod
    def camel_to_snake(text: str) -> str:
        """Convert camelCase to snake_case (Python convention)"""
        import re
        return re.sub(r'(?<!^)(?=[A-Z])', '_', text).lower()
    
    @staticmethod
    def snake_to_camel(text: str) -> str:
        """Convert snake_case to camelCase"""
        components = text.split('_')
        return components[0] + ''.join(x.title() for x in components[1:])


@contextmanager
def measure_performance(operation: str):
    """
    Context manager for measuring performance
    Demonstrates: Context manager decorator, Python idioms
    """
    print(f"[PERF] Starting: {operation}")
    start = time.time()
    try:
        yield
    finally:
        elapsed = time.time() - start
        print(f"[PERF] Completed: {operation} in {elapsed:.6f}s")


def demonstrate_cogutil():
    """Main demonstration function"""
    print("=" * 70)
    print("OpenCog Cogutil - Python Utility Library Demo")
    print("Showcasing Python's strengths: Dynamic typing, rich stdlib, elegance")
    print("=" * 70)
    print()
    
    # Logger demonstration
    print("1. Logger Demonstration")
    print("-" * 50)
    logger = Logger("CogUtil", logging.INFO)
    logger.info("Cogutil library initialized")
    logger.debug("This debug message won't show (level too low)")
    logger.warn("This is a warning message")
    logger.error("This is an error message")
    
    logger.set_level(logging.DEBUG)
    logger.debug("Now debug messages are visible")
    print()
    
    # Config demonstration
    print("2. Configuration Manager")
    print("-" * 50)
    config = Config()
    config.set("opencog.version", "1.0.0")
    config["atomspace.enabled"] = "true"  # Dictionary-style access
    config["cogserver.port"] = "17001"
    
    logger.info("Configuration loaded:")
    config.dump()
    print()
    
    logger.info(f"Port setting: {config.get('cogserver.port')}")
    print()
    
    # Timer demonstration with context manager
    print("3. Timer Context Manager")
    print("-" * 50)
    with Timer("Processing", logger) as timer:
        logger.info("Simulating some work...")
        # Simulate work
        total = sum(range(1_000_000))
    print()
    
    # Performance measurement context manager
    print("4. Performance Context Manager")
    print("-" * 50)
    with measure_performance("Data processing"):
        time.sleep(0.01)  # Simulate work
    print()
    
    # String utilities demonstration
    print("5. String Utilities")
    print("-" * 50)
    logger.info("String utilities demonstration:")
    text = "OpenCog,AtomSpace,CogServer,Cogutil"
    parts = StringUtils.split(text, ',')
    
    logger.info("Split result:")
    for part in parts:
        print(f"  - {part}")
    
    joined = StringUtils.join(parts, " + ")
    logger.info(f"Joined: {joined}")
    
    logger.info(f"Uppercase: {StringUtils.to_upper('opencog rocks')}")
    logger.info(f"Lowercase: {StringUtils.to_lower('OPENCOG ROCKS')}")
    logger.info(f"Trimmed: '{StringUtils.trim('  spaced out  ')}'")
    
    # Python-specific: Case conversions
    logger.info(f"camelCase → snake_case: {StringUtils.camel_to_snake('myVariableName')}")
    logger.info(f"snake_case → camelCase: {StringUtils.snake_to_camel('my_variable_name')}")
    print()
    
    # Decorator demonstration
    print("6. Function Decorator")
    print("-" * 50)
    
    @timeit
    def fibonacci(n: int) -> int:
        """Calculate Fibonacci number"""
        if n <= 1:
            return n
        return fibonacci(n-1) + fibonacci(n-2)
    
    result = fibonacci(20)
    logger.info(f"Fibonacci(20) = {result}")
    print()
    
    # List comprehension demonstration
    print("7. List Comprehensions (Pythonic)")
    print("-" * 50)
    squares = [x**2 for x in range(10)]
    logger.info(f"Squares: {squares}")
    
    evens = [x for x in range(20) if x % 2 == 0]
    logger.info(f"Even numbers: {evens}")
    print()
    
    logger.info("Cogutil demonstration complete!")
    print("=" * 70)
    print("Python strengths demonstrated:")
    print("  ✓ Clean, readable syntax")
    print("  ✓ Rich standard library (logging, time, etc.)")
    print("  ✓ Context managers for resource management")
    print("  ✓ Decorators for code elegance")
    print("  ✓ List comprehensions for conciseness")
    print("  ✓ Duck typing and dynamic behavior")
    print("=" * 70)


if __name__ == "__main__":
    demonstrate_cogutil()
